### PR TITLE
Create Pod Functionality

### PR DESCRIPTION
**Pull request by strzinek**
- Allow to create new pods with port mappings - new kebab control on containers panel
- Show pod owner in listings - under pod name
- Correctly limit container ownership when creating container in existing pod - select owner according to pod and disable owner selection

### CODE REVIEW

**Review by jelly — CHANGES_REQUESTED** [review]
Thanks for the PR! This was something we wanted to add for a while!

@garrett as we have limited set of integrations, do we still want to have a tabbed layout?

image

I've noticed that pods also supports volumes/mounts so should we add that as well?

File: src/Containers.jsx [hunk 1]
```diff
@@ -523,6 +525,24 @@ class Containers extends React.Component {
...
+                          isOpen={isActionsKebabOpen}
+                          isPlain
+                          position="right"
+                          dropdownItems={[
+                              <DropdownItem key="create-new-pod"
+                                  component="button"
+                                  onClick={() => this.setState({ showCreatePodModal: true })}>
+                                  {_("Create pod")}
+                              </DropdownItem>
+                          ]} />
```
[the remark this answers is not in the document]
Comment (strzinek) [reply src/Containers.jsx:558]: I am planning to implement other functions later which could go to kebab menu, and which are available through the API, e.g. Prune unused pods, Play a kube YAML, ... And I think two buttons together with filters would not look good on narrow layouts too. What do you think?

File: src/Containers.jsx [hunk 2]
```diff
@@ -378,13 +378,22 @@ class Containers extends React.Component {
...
+                { title: <Badge isRead className={containerStateClass}>{_(containerState)}</Badge> }, // States are defined in util.js
+            ];
+        else
+            columns = [
+                { title: info_block },
+                { title: container.isSystem ? _("system") : <div><span className="ct-grey-text">{_("user:")} </span>{this.props.user}</div>, props: { modifier: "nowrap" } },
+                { title: proc, props: { modifier: "nowrap" } },
+                { title: mem, props: { modifier: "nowrap" } },
+                { title: <Badge isRead className={containerStateClass}>{_(containerState)}</Badge> }, // States are defined in util.js
+            ];
```
[the remark this answers is not in the document]
Comment (garrett) [reply src/Containers.jsx:396]: Port mappings an mounts are at the pod level, right? I thought I had a design for this somewhere. (But perhaps I'm thinking of something else...?)

I'll look around, but it's end of the day today, so please ping me if I forget and take a while to get back to this.

[the remark this answers is not in the document]
Comment (garrett) [reply src/Containers.jsx:396]: I wasn't able to find what I was thinking of. Perhaps I have a pattern in mind that was in another mockup.

This means I'll need to mock up what I have in mind for Cockpit-Podman. :wink:

**Comment by garrett** [comment]
> @garrett as we have limited set of integrations, do we still want to have a tabbed layout?

No, please drop the tabs from this modal dialog. Thanks!

File: src/PodCreateModal.jsx [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    componentDidMount() {
+        this._isMounted = true;
+    }
+
+    componentWillUnmount() {
+        this._isMounted = false;
+
+        if (this.activeConnection)
+            this.activeConnection.close();
+    }
```
[the remark this answers is not in the document]
Comment (strzinek) [reply src/PodCreateModal.jsx]: Removed

File: src/PodCreateModal.jsx [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+import * as client from './client.js';
+import cockpit from 'cockpit';
+
+import "./PodCreateModal.scss";
+
+const _ = cockpit.gettext;
+
+const systemOwner = "system";
+
+const PublishPort = ({ id, item, onChange, idx, removeitem, itemCount }) =>
```
[the remark this answers is not in the document]
Comment (strzinek) [reply src/PodCreateModal.jsx]: Moved to separate files.

File: src/PodCreateModal.jsx [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+                        const pm = { container_port: parseInt(port.containerPort), protocol: port.protocol };
+                        if (port.hostPort !== null)
+                            pm.host_port = parseInt(port.hostPort);
+                        if (port.IP !== null)
+                            pm.host_ip = port.IP;
+                        return pm;
+                    });
+
+        return createConfig;
+    }
```
[the remark this answers is not in the document]
Comment (strzinek) [reply src/PodCreateModal.jsx]: Pod config parameters differ from container.

**Comment by strzinek** [comment]
Tabs removed.

I also added volumes section, but found out that mounts wont get created for pod because of bug in Podman: https://github.com/containers/podman/issues/13548 . I am not able to verify it is fixed in 4.x version as I am on fedora 35 for now.

I would also add a new test for creating pod later on.

**Comment by strzinek** [comment]
Now it is possible to choose if infra container is created. Also, infra containers are not hidden in pods listings anymore.

I also prepared two new tests, TestApplication.testCreatePodUser and TestApplication.testCreatePodSystem.

File: src/Containers.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (jelly) [reply src/Containers.jsx:396]: @garrett the pod details design is here https://github.com/cockpit-project/cockpit-podman/issues/489 but that does not deal with the owner field. Maybe we can improve this in a follow up PR.

**Review by jelly — CHANGES_REQUESTED** [review]

**Comment by jelly** [comment]
I think this is a good PR to look at landing next after the rename PR.

File: src/Containers.jsx [hunk 6]
```diff
@@ -463,9 +484,6 @@ class Containers extends React.Component {
                 );
             }
 
-            // Remove infra containers
-            filtered = filtered.filter(id => !this.props.containers[id].IsInfra);
```
[the remark this answers is not in the document]
Comment (strzinek) [reply src/Containers.jsx:467]: Infra container is a placeholder for shared resources of pod containers. The pod port mapping for example is defined on it. Without showing infra container, user of cockpit-podman does not know if pod has such shared resources and what ports pod exposes.

**Comment by jelly** [comment]
The rename PR, can you rebase the PR please?

**Comment by jelly** [comment]
I've rebased the PR, resolved the conflicts and dropped some of the things which can be a follow up PR.

**Comment by marusak** [comment]
This breaks pixel tests: https://cockpit-logs.us-east-1.linodeobjects.com/pull-961-20220819-084819-8540cbeb-fedora-35/TestApplication-testRunImageSystem-integration-pixels.png

It causes regression, I did this change a few weeks ago - wrong checkout?

Also needs some new tests

**Comment by jelly** [comment]
> This breaks pixel tests: https://cockpit-logs.us-east-1.linodeobjects.com/pull-961-20220819-084819-8540cbeb-fedora-35/TestApplication-testRunImageSystem-integration-pixels.png
> 
> It causes regression, I did this change a few weeks ago - wrong checkout?
> 
> Also needs some new tests

Hmm the code was moved around so maybe it doesn't pick up some css changes?

**Comment by jelly** [comment]
Failing tests on ubuntu-2204 are due to podman not support user mounts on 3.4.4 and requires 4.0.0 so our gui should hide it https://github.com/containers/podman/issues/10379

**Comment by jelly** [comment]
Issue summary:

* Fedora-coreos/debian-testing testCreatePodUser fails for unknown reasons
* Changed Pixel tests for https://cockpit-logs.us-east-1.linodeobjects.com/pull-961-20220927-091133-81d25dcc-fedora-36/pixeldiff.html#TestApplication-testRunImageSystem-integration-medium
* Requires a new review

**Comment by marusak** [comment]
Why `Owner` label is not on the same row?
Screenshot from 2022-09-29 13-52-30

Like here for example?
Screenshot from 2022-09-29 13-52-18

Also found it a bit confusing, that when you create pod and give it ports and volumes, you don't see them in the UI unless you add there container and start it.

**Comment by jelly** [comment]
> Also found it a bit confusing, that when you create pod and give it ports and volumes, you don't see them in the UI unless you add there container and start it.

Agreed, however.... the Ports can be found in `infraContainer` but for the mounts we need `infraContainerDetails` and that is only available when the pod is started. We should ask the podman guys to expose the mounts in the containers information to make this work nicely.

**Review by marusak — APPROVED** [review]
Superb, thank you!

**Comment by marusak** [comment]
Can you please write proper release note with screenshot as well?

**Review by marusak — APPROVED** [review]

**Comment by strzinek** [comment]
Thanks everyone for pushing things through, I haven't had any time to pursue this myself.

**Comment by jelly** [comment]
> Thanks everyone for pushing things through, I haven't had any time to pursue this myself.

No problem, thanks for the initial PR!

**Pull request merged**